src/runtime_src/core/common/module_loader.h

```diff
@@ -0,0 +1,85 @@
+/**
+ * Copyright (C) 2016-2020 Xilinx, Inc
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"). You may
+ * not use this file except in compliance with the License. A copy of the
+ * License is located at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ */
+
+#ifndef xrtcore_util_module_loader_h_
+#define xrtcore_util_module_loader_h_
+
+// This file contains a loader utility class for plugin modules
+//  that are loaded from either OpenCL or XRT level applications.
+
+#include <boost/filesystem/path.hpp>
+#include <functional>
+#include <string>
+
+#include "core/common/config.h"
+
+namespace xrt_core {
+
+  /*
+   * class: module_loader
+   * 
+   * This class is responsible for loading a plugin module from the
+   * appropriate directory under the XILINX_XRT installation.  The loading
+   * happens at object construction time, so the XRT side implementation should
+   * contain a function that instantiates a single static instance of this 
+   * class to handle the loading of a module once in a thread safe manner.
+   *
+   */
+
+  class module_loader
```

Comment:
Why is this a class?

Reply:
The loading of the plugin module should occur only once and is triggered by the first call into XRT source code from the user's side.  All of the entry points specific to a plugin will call a single function to load that plugin module.  In order to handle multithreaded access from the user and guarantee that the loading only occurs once, the loading functions consist of a single static declaration of an instance of this class.  This instance is only constructed once and is thread safe per C++-11, and simplifies the code by not having to create global locks for each plugin and check flags on every callback to see if loading should occur.

Reply:
ok. makes sense